### PR TITLE
Feature/history per document

### PR DESCRIPTION
**Pull request by iNewLegend**

### CODE REVIEW

**Review by matipojo — CHANGES_REQUESTED** [review]

File: assets/dev/js/modules/modules.js [hunk 1]
```diff
@@ -14,3 +14,5 @@ export default window.elementorModules = {
 		Masonry: Masonry,
 	},
 };
+
+export default window.elementorModules = elementorModules;
```
Comment (matipojo) [inline assets/dev/js/modules/modules.js]: Why?

File: modules/history/assets/js/panel-page.js [hunk 2]
```diff
@@ -19,25 +19,35 @@ module.exports = Marionette.LayoutView.extend( {
 
 	currentTab: null,
 
-	initialize: function() {
+	documentId: null,
```
Comment (matipojo) [inline modules/history/assets/js/panel-page.js]: It's used?

File: modules/history/assets/js/revisions/panel-tab.js [hunk 3]
```diff
@@ -34,9 +38,10 @@ module.exports = Marionette.CompositeView.extend( {
 	getRevisionViewData: function( revisionView ) {
 		var self = this;
 
-		elementor.history.revisions.getRevisionDataAsync( revisionView.model.get( 'id' ), {
+		this.document.revisions.getRevisionDataAsync( revisionView.model.get( 'id' ), {
 			success: function( data ) {
-				elementor.history.revisions.setEditorData( data.elements );
+				self.document.revisions.setEditorData( data.elements );
```
Comment (matipojo) [inline modules/history/assets/js/revisions/panel-tab.js]: remove all `self` from file

File: modules/history/assets/js/revisions/panel-tab.js [hunk 4]
```diff
@@ -97,21 +123,21 @@ module.exports = Marionette.CompositeView.extend( {
...
-		elementor.history.history.getItems().reset();
+		this.document.history.getItems().reset();
 	},
 
 	onDiscardClick: function() {
-		elementor.history.revisions.setEditorData( elementor.config.data );
+		this.document.revisions.setEditorData( elementor.config.data );
 
-		elementor.saver.setFlagEditorChange( this.isRevisionApplied );
+		$e.run( 'document/save/set-is-modified', { status: this.isRevisionApplied } );
```
Comment (matipojo) [inline modules/history/assets/js/revisions/panel-tab.js]: Revert

File: modules/history/assets/js/revisions/panel-tab.js [hunk 5]
```diff
@@ -97,21 +123,21 @@ module.exports = Marionette.CompositeView.extend( {
 	},
 
 	onApplyClick: function() {
-		elementor.saver.setFlagEditorChange( true );
+		$e.run( 'document/save/set-is-modified', { status: true } );
```
Comment (matipojo) [inline modules/history/assets/js/revisions/panel-tab.js]: Revert

File: modules/history/assets/js/revisions/panel-tab.js [hunk 6]
```diff
@@ -161,7 +187,7 @@ module.exports = Marionette.CompositeView.extend( {
 		childView.$el.addClass( 'elementor-revision-current-preview elementor-revision-item-loading' );
 
 		if ( elementor.saver.isEditorChanged() && ( null === self.currentPreviewId || elementor.config.current_revision_id === self.currentPreviewId ) ) {
-			elementor.saver.saveEditor( {
+			$e.run( 'document/save/save', {
```
Comment (matipojo) [inline modules/history/assets/js/revisions/panel-tab.js]: Revert

File: tests/qunit/core/editor/document/elements/commands/delete.spec.js [hunk 7]
```diff
@@ -15,7 +15,7 @@ export const Delete = () => {
 				assert.equal( eColumn.view.collection.length, 1, 'Button #1 were deleted.' );
 
 				// Ensure editor saver.
-				elementor.saver.setFlagEditorChange( false );
+				$e.run( 'document/save/set-is-modified', { status: false } );
```
Comment (matipojo) [inline tests/qunit/core/editor/document/elements/commands/delete.spec.js]: Revert

File: tests/qunit/core/editor/document/elements/commands/empty.spec.js [hunk 8]
```diff
@@ -10,7 +10,7 @@ export const Empty = () => {
 			ElementsHelper.createButton( eColumn );
 
 			// Ensure editor saver.
-			elementor.saver.setFlagEditorChange( false );
+			$e.run( 'document/save/set-is-modified', { status: false } );
```
Comment (matipojo) [inline tests/qunit/core/editor/document/elements/commands/empty.spec.js]: Revert

File: tests/qunit/core/editor/document/elements/commands/paste-style.spec.js [hunk 9]
```diff
@@ -12,7 +12,7 @@ export const PasteStyle = () => {
 				ElementsHelper.copy( eButtonStyled );
 
 				// Ensure editor saver.
-				elementor.saver.setFlagEditorChange( false );
+				$e.run( 'document/save/set-is-modified', { status: false } );
```
Comment (matipojo) [inline tests/qunit/core/editor/document/elements/commands/paste-style.spec.js]: Revert

File: tests/qunit/core/editor/document/elements/commands/paste.spec.js [hunk 10]
```diff
@@ -12,7 +12,7 @@ export const Paste = () => {
 				ElementsHelper.copy( eButton );
 
 				// Ensure editor saver.
-				elementor.saver.setFlagEditorChange( false );
+				$e.run( 'document/save/set-is-modified', { status: false } );
```
Comment (matipojo) [inline tests/qunit/core/editor/document/elements/commands/paste.spec.js]: Revert

File: tests/qunit/core/editor/document/elements/commands/reset-style.spec.js [hunk 11]
```diff
@@ -8,7 +8,7 @@ export const ResetStyle = () => {
 				const eButtonStyled = ElementsHelper.createAutoButtonStyled();
 
 				// Ensure editor saver.
-				elementor.saver.setFlagEditorChange( false );
+				$e.run( 'document/save/set-is-modified', { status: false } );
```
Comment (matipojo) [inline tests/qunit/core/editor/document/elements/commands/reset-style.spec.js]: Revert

File: tests/qunit/core/editor/document/history/component.spec.js [hunk 12]
```diff
@@ -86,7 +86,7 @@ jQuery( () => {
 
 		// TODO: Temp next tests are not in the right place.
 		QUnit.test( 'Saver Editor Flag', ( assert ) => {
-			elementor.saver.setFlagEditorChange( false );
+			$e.run( 'document/save/set-is-modified', { status: false } );
```
Comment (matipojo) [inline tests/qunit/core/editor/document/history/component.spec.js]: Revert

File: tests/qunit/core/editor/document/manager.js [hunk 13]
```diff
@@ -0,0 +1,51 @@
+import ElementsHelper from './helper';
```
Comment (matipojo) [inline tests/qunit/core/editor/document/manager.js]: Rename `spec.js`

**Review by matipojo — CHANGES_REQUESTED** [review]

File: assets/dev/js/editor/components/settings/page/manager.js [hunk 14]
```diff
@@ -19,20 +19,23 @@ module.exports = BaseSettings.extend( {
...
 		template: function() {
-			elementor.saver.saveAutoSave( {
-				onSuccess: function() {
-					elementor.reloadPreview();
-
-					elementor.once( 'preview:loaded', function() {
-						$e.route( 'panel/page-settings/settings' );
-					} );
+			$e.run( 'document/save/auto', {
+				force: true,
```
[the remark this answers is not in the document]
Comment (iNewLegend) [reply assets/dev/js/editor/components/settings/page/manager.js]: Old page/manager:
```javascript
		template: function() {
			elementor.saver.saveAutoSave( {
				onSuccess: function() {
					elementor.reloadPreview();

					elementor.once( 'preview:loaded', function() {
						$e.route( 'panel/page-settings/settings' );
					} );
				},
			} );
		},
```
Old saver/manager:
```javascript
	doAutoSave: function() {
		var editorMode = elementor.channels.dataEditMode.request( 'activeMode' );

		// Avoid auto save for Revisions Preview changes.
		if ( 'edit' !== editorMode ) {
			return;
		}

		this.saveAutoSave();
	},

	saveAutoSave: function( options ) {
		if ( ! this.isEditorChanged() ) {
			return;
		}

		options = _.extend( {
			status: 'autosave',
		}, options );

		this.saveEditor( options );
	},
```

new auto save command:
```javascript
import Base from './base/base';

export class Auto extends Base {
	apply( args ) {
		const { force = false, options = {} } = args;

		if ( ! force && 'edit' !== elementor.channels.dataEditMode.request( 'activeMode' ) ) {
			return;
		}

		this.save( options );
	}

	save( options ) {
		if ( this.component.isEditorChanged() ) {
			// TODO: Move to es6.
			options = _.extend( {
				status: 'autosave',
			}, options );

			$e.run( 'document/save/save', options );
		}
	}
}

export default Auto;
```

File: assets/dev/js/editor/editor-base.js [hunk 15]
```diff
@@ -1,43 +1,51 @@
-/* global ElementorConfig */
```
[the remark this answers is not in the document]
Comment (iNewLegend) [reply assets/dev/js/editor/editor-base.js]: ES6 editor base merged.

File: assets/dev/js/editor/editor-base.js [hunk 16]
```diff
@@ -1,43 +1,51 @@
...
+import Environment from 'elementor-common/utils/environment';
```
[the remark this answers is not in the document]
Comment (iNewLegend) [reply assets/dev/js/editor/editor-base.js]: ES6 editor base merged.

File: assets/dev/js/modules/modules.js (hunk 1 above)
Comment (iNewLegend) [reply assets/dev/js/modules/modules.js]: Reverted.

**Review by matipojo — DISMISSED** [review]

File: assets/dev/js/editor/document/save/commands/discard.js [hunk 17]
```diff
@@ -0,0 +1,14 @@
+import Base from './base/base';
+
+export class Discard extends Base {
+	apply() {
+		elementorCommon.ajax.addRequest( 'discard_changes', {
+			success: function() {
```
Comment (kobizz) [inline assets/dev/js/editor/document/save/commands/discard.js]: Use arrow function

File: assets/dev/js/editor/document/save/commands/pending.js [hunk 18]
```diff
@@ -0,0 +1,16 @@
+import Base from './base/base';
+
+export class Pending extends Base {
+	apply( args ) {
+		let { options } = args;
+
+		// TODO: Move to es6.
```
Comment (kobizz) [inline assets/dev/js/editor/document/save/commands/pending.js]: Do it

File: assets/dev/js/editor/document/save/commands/publish.js [hunk 19]
```diff
@@ -0,0 +1,15 @@
+import Base from './base/base';
+
+export class Publish extends Base {
+	apply( args ) {
+		let { options = {} } = args;
+
+		// TODO: Move to es6.
```
Comment (kobizz) [inline assets/dev/js/editor/document/save/commands/publish.js]: Do it

File: assets/dev/js/editor/document/save/commands/update.js [hunk 20]
```diff
@@ -0,0 +1,15 @@
+import Base from './base/base';
+
+export class Update extends Base {
+	apply( args ) {
+		let { options = {} } = args;
+
+		// TODO: Move to es6.
```
Comment (kobizz) [inline assets/dev/js/editor/document/save/commands/update.js]: Do it

File: assets/dev/js/editor/document/save/component.js [hunk 21]
```diff
@@ -0,0 +1,165 @@
...
+			update: ( args ) => ( new Commands.Update( args ).run() ),
+		};
+	}
+
+	saveEditor( options ) {
+		if ( this.isSaving ) {
+			return;
+		}
+
+		// TODO: Move to es6.
```
Comment (kobizz) [inline assets/dev/js/editor/document/save/component.js]: Do it

**Review by kobizz — CHANGES_REQUESTED** [review]

**Review by kobizz — APPROVED** [review]

**Pull request merged**